File: fenwickTree.go
// https://tobin.cc/blog/fenwick/

package leetcode

type Fenwick struct {
	tree []int
}

// NewFenwick: Build Fenwick tree to hold n values.
func NewFenwick(n int) *Fenwick {
	fen := &Fenwick{
		tree: make([]int, n+1),
	}
	return fen
}

// Sum all values upto and including index.
func (fen *Fenwick) Sum(index int) int {
	sum := 0
	index++ // fenwick tree is 1-based
	for index > 0 {
		sum += fen.tree[index]
		index -= lsb(index)
	}
	return sum
}

// lsb: Least Significant Bit
func lsb(x int) int {
	return x & -x
}

// Update using addition index by value.
func (fen *Fenwick) Update(index, delta int) {
	index++ // fenwick tree is 1-based
	for index <= fen.Size() {
		fen.tree[index] += delta
		index += lsb(index)
	}
}

// Size: Number of values stored by tree.
func (fen *Fenwick) Size() int {
	return len(fen.tree)
}

File: fenwickTree.py
class Bit:
    def __init__(self, op, e, n):
        self.op = op
        self.e = e
        self.n = n+1
        self.table = [e() for _ in range(self.n)]

    def update(self, i, x):
        i += 1
        while i < self.n:
            self.table[i] = self.op(self.table[i], x)
            i += i & -i

    def prod(self, i):
        i += 1
        res = self.e()
        while i > 0:
            res = self.op(res,self.table[i])
            i -= i & -i
        return res







class Fenwick:
    def __init__(self, n):
        self.a = [0] * (n + 1)

    def sum(self, i):
        s = 0
        while i > 0:
            s += self.a[i]
            i -= i & -i
        return s

    def add(self, i, x):
        while i < len(self.a):
            self.a[i] += x
            i += i & -i


